### ISSUE TITLE
change the effect on an element runtime

### ISSUE THREAD

**Issue by kinotto** [jlmakes/scrollreveal.js#301]
Hello and thanks for this great plugin,
i read carefully your documentation and i cannot find a way to change runtime an effect,  let's take my plnkr  as an example (attached at the end of the page), i have a simple div on which i'm applying on load an effect, this works fine, but if i try to change it for example by triggering a click on a button it doesn't works. 
it works only if i took it out of the viewport and then take it back again (just by scrolling up and down the page)

you can find my example here,
https://plnkr.co/edit/BOKXeu4HwN5bcJGkmP3q?p=preview

in the example i'm just changing the field 'origin' from 'bottom' to 'top'.

As i said i'm looking for a way  maybe i'm just missing something, any help would be much appreciated .

thanks for your work again.

**Comment by jlmakes** [comment]
So I noticed a few things.

### 1.
ScrollReveal will not work properly when passed a jQuery object, so your code here is invalid:
```js
sr.reveal($('.foo'), { origin: 'top', delay: 500, scale: 0.9, reset: true });
```
Instead, you should pass either the `.foo` selector alone, or a DOM element. What I've done instead, is store a reference to a DOM element (for performance and ease of use), and used that throughout the code.

For example:
```js
var foo = document.querySelector('.foo');

sr.reveal(foo, {
  origin: 'top',
  delay: 500,
  scale: 0.9,
  reset: true
});
```

### 2.

The element already has styles applied appended to the `style` attribute, so if you want the animation to _replay_ without leaving/re-entering the viewport, you'll need to remove the current `style` attribute.

I have taken this a step farther, and instead of removing the `style` attribute, I have instead set it's value to the original inline styles ScrollReveal stored before appending anything (using the `data-sr-id` attribute to grab the correct store key).
```js
var elemId = foo.getAttribute('data-sr-id');
var styles = sr.store.elements[elemId].styles.inline;
  
foo.setAttribute('style', styles);
```

### 3.

ScrollReveal checks for `data-sr-id` attributes to deduce whether or not an element has been registered during the setup process. In other words, if you want ScrollReveal to re-reveal this element like it's brand new, you'll need to remove the `data-sr-id` attribute as well.
```js
foo.removeAttribute('data-sr-id');
```

### Putting It Together
```js
var foo = document.querySelector('.foo');

window.sr = ScrollReveal();

var fooReveal = {
  origin: 'bottom',
  delay: 200,
  distance: '90px',
  easing: 'ease-in-out',
  rotate: {
    z: 10
  },
  scale: 1.1,
  reset: true
};

sr.reveal(foo, fooReveal);

var fn = function() {

  var elemId = foo.getAttribute('data-sr-id');
  var styles = sr.store.elements[elemId].styles.inline;
  
  foo.setAttribute('style', styles);
  foo.removeAttribute('data-sr-id');

  sr.reveal(foo, {
    origin: 'top',
    delay: 500,
    scale: 0.9,
    reset: true
  });
}
```

### [See It Action](https://plnkr.co/edit/cVvEaBsMQcV0U4G4QwXF?p=preview)

Hope this helps!

**Comment by kinotto** [comment]
that is exactly what i was looking for,
in my humble opinion the points **.2** and  **.3** should be mentioned in the documentation
thanks again :)

**Comment by jlmakes** [comment]
No problem @kinotto and I agree, this aspect of ScrollReveal deserves attention.

Feature request #227 proposes a new public method -- e.g. `sr.clean()` -- to remove styles and event listeners associated with an element set -- which could include points **2** and **3**.

Meaning your code snippet could look like this:
```js
var fn = function() {
  sr.clean('.foo').reveal('.foo', {
    origin: 'top',
    delay: 500,
    scale: 0.9,
    reset: true
  });
}
```

**Comment by kinotto** [comment]
Hi,
i need your help again,
i'm trying to remove the scroll reveal after is applied to a a set of elements of the DOM
These set of elements is retrieved by class.


i have succesfully applied the "replay" effect that we discuss few days ago, but i'm not capable to remove completely the effect and then applying it back again.
I'm getting this exception:

`scrollreveal.js:732 Uncaught TypeError: Cannot read property 'domEl' of undefined
`

i've created a clear example of the error here:
https://plnkr.co/edit/BOKXeu4HwN5bcJGkmP3q?p=preview

Basically the error occured when you remove the effect then apply it back.
following your advices , i was thinking that removing the element from the store map would have been enough but this is not the case.

  `delete sr.store.elements[elemId];`

i think that this instruction is messing everything up


Hope you can help me again, have a great day.

**Comment by jlmakes** [comment]
Your `remove()` function deletes the `data-sr-id` among other things, which means when you call `replay()` your function fails because it relies on things just deleted.

**Comment by kinotto** [comment]
I think i did not explain my self right, actually is not that easy.
I'm working on a big content management system and i want to apply these kind of effects to, let's say, some boxes, i need two functionalities:
- replay an effect on click
- remove an effect on click, i want to erase every reference from the library.

i was able to implement with your help the replay functionality, but i'm still unable to implement the removal.

i've just tried to adjust the plnkr to my original code.
https://plnkr.co/edit/BOKXeu4HwN5bcJGkmP3q?p=preview


If you replay the effect it works, but if you do a replay followed by a remove it doesn't, the scroll reveal it's still there infact if you "hide" the boxes from the viewport through a scroll for example and then put them back in you will see again the reveal, it seems that the library keep a reference to the elements beside the `data-sr-id` field, but maybe i'm missing something.

the exception that you underlined in the previous post was clearly an error, but i'm facing a different problem here. thanks again for your help :)

**Comment by jlmakes** [comment]
Please, this issue arises out of your highly specific use case, and isn't supported by the public API.

The `_animate()` method is called on scroll and resize events, which scans through `sr.store.elements` -- which is why you still see animations after `remove()`.  One solution could be to remove the event listeners, such as outlined in #227 mentioned above.

That said, I'm closing this discussion because:
- We're fighting against ScrollReveal's behavior for an use case you have in mind
- Official support for removing ScrollReveal will be introduced soon.
